phylo/src/optimisers/topo_optimiser.rs

```diff
@@ -0,0 +1,122 @@
+use log::{debug, info};
+
+use crate::likelihood::PhyloCostFunction;
+use crate::optimisers::{BranchOptimiser, PhyloOptimisationResult, PhyloOptimiser};
+use crate::phylo_info::PhyloInfo;
+use crate::tree::{NodeIdx, Tree};
+use crate::Result;
+
+pub struct TopologyOptimiser<'a, EM: PhyloCostFunction> {
+    pub(crate) epsilon: f64,
+    pub(crate) model: &'a EM,
+    pub(crate) info: PhyloInfo,
+}
+
+impl<'a, EM: PhyloCostFunction + Clone> PhyloOptimiser<'a, EM> for TopologyOptimiser<'a, EM> {
+    fn new(model: &'a EM, info: &PhyloInfo) -> Self {
+        TopologyOptimiser {
+            epsilon: 1e-3,
+            model,
+            info: info.clone(),
+        }
+    }
+
+    fn run(self) -> Result<PhyloOptimisationResult> {
+        debug_assert!(self.info.tree.len() > 3);
+
+        info!("Optimising tree topology with SPRs.");
+        let mut info = self.info.clone();
+        let initial_logl = self.model.cost(&info, true);
+
+        info!("Initial logl: {}.", initial_logl);
+        let mut curr_cost = initial_logl;
+        let mut prev_cost = f64::NEG_INFINITY;
+        let mut iterations = 0;
+
+        // No pruning on the root branch
+        let prune_locations: Vec<NodeIdx> = info
+            .tree
+            .preorder()
+            .iter()
+            .filter(|&n| n != &info.tree.root)
+            .cloned()
+            .collect();
+        while (curr_cost - prev_cost) > self.epsilon {
+            iterations += 1;
+            info!("Iteration: {}.", iterations);
+            prev_cost = curr_cost;
+            info!("Current logl: {}.", curr_cost);
+            for prune_branch in &prune_locations {
+                if info.tree.children(&info.tree.root).contains(prune_branch) {
+                    // due to topology change the current node may have become the direct child of root
+                    continue;
+                }
+                let regraft_locations = Self::find_regraft_options(prune_branch, &info);
+                let mut moves: Vec<(f64, Tree)> = Vec::with_capacity(regraft_locations.len());
+                for regraft_branch in &regraft_locations {
+                    let mut new_info = info.clone();
+                    new_info.tree = info.tree.rooted_spr(prune_branch, regraft_branch).unwrap();
```

Comment:
nit: something feels wrong here. is it certain this place cannot fail? I'd recommend error handling. probably ? should be better than just `unwrap`

Reply:
also it's a strange `clone()` here. I would recommend moving this type of "cloning" to the tree inteself. e.g. `clone_for_tree`

Reply:
This should never fail, but yes, ? is definitely better than unwrap().

Reply:
agreed for the second point, I am cloning a big structure that I don't need to clone at all. I will refactor that.

Reply:
This will be refactored in the next PR as per our discussion